homeassistant/components/suez_water/coordinator.py

```diff
@@ -48,61 +34,38 @@ def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
         )
 
     async def _async_setup(self) -> None:
-        self._sync_client = await self.hass.async_add_executor_job(self._get_client)
+        self._async_client = await self._get_client()
+        self._data_api = SuezData(self._async_client)
```

Comment:
I'd personally make the client have a bit more state and just have every class in the client instead of creating a second object imo

Reply:
Personally I like to split objects responsibilities but yes I'll do it in the library and expose only one endpoint here.

Reply:
Yeah I understand what you mean. It's clear for me but should be for everyone.
I'll clean it up 😅

Reply:
Oh and don't forget, this was a personal opinion, but I do like to be opinionated with things like this to help get to a nicer output :)

Reply:
Don't worry I understand what you mean. Most of my day job is reviewing PR and I'm pretty opinionated also.

And you're comment made me realize I exposed (and used) something that should be internal.

Reply:
https://github.com/joostlek/python-spotify, this is how I usually have my libraries :)

Reply:
Oh thanks I'll have a look, and make some changes, probably not all of them here 😄 .

Reply:
Ok, I gave a quick look at your library.
I understand pretty well how you did structure it.

I still don't think everything should be in the same file or class. It might be because I'm a Java / TS person but I do like the fact that the responsibilities are splitted between multiple services, for examples in my case :
- a technical client, responsible only of the technical part, establishing session, making requests, ...
- a data client, consuming the technical one and not knowing of how requests are made

Currently I agree that the separation is not clear enough and that it also should be invisible when using it.

But as I'm still very young to python and more so to HA, if in your opinion a python library shouldn't be structured that way, I'll gladly follow allong.

Reply:
I mean there is not a strict way. I'm also coming from a java/kotlin and typescript background, and in that world I kinda do see why you structure code that way, since those projects are usually a tad larger.

But in this case, the library is a tiny cog in a huge machine (that only touches a small part). So when I take your argument about being able to change how I fetch data in my libraries, in theory i could change the implementation of `_request`, to use a different library like `httpx`, i can change that without changing the functions that tell it what it needs.

So is it a real separation of concerns? Nope. Is it real object oriented? Nope (but that also got slapped onto Python in a later stage).

But I do think a lot of power in python lies in keeping it simple and stupid. If you have only 1 way of fetching data, why give users the option? (Unless you are going to add clients for idk, `requests`, `aiohttp` and `httpx`, but if you don't, why bother?)

But to end it, don't forget to do what feels good. We as an open source project should not force structures onto maintainers, as the library is usually the freedom they have, as the code in the integration is thoroughly reviewed. So the structure in my libraries works fine (I copied it from Frenck) and it works well for some other people and their use cases, but if it doesn't for you or your use cases, feel free to find your own way. (We do still check the code for common mistakes regarding either asyncio or incorrect python usage or obvious bugs).

Oh this turned out to be a book. If you ever get around making a discord account and want to discuss more, be sure to hmu ;)

Reply:
Loved your book 😆

Yeah, it make a lot of sense, this library is tiny so keeping it simple is a great idea.

Too much deformation over the years 😅
I might start loving this lack of rigidity in the language.

> 
> Oh this turned out to be a book. If you ever get around making a discord account and want to discuss more, be sure to hmu ;)
I did create an account just didn't use it beside joining HA server.

Reply:
The best thing about python Vs java/kotlin/typescript is that the way that you write code actually has an impact on the speed and efficiency. Although with the other languages that is caught by a compiler or transpiler, here it's our duty to make sure it's nice